### DISCUSSION TITLE
pipeline mode (libpq's pipeline API)

### DISCUSSION THREAD

**Discussion by karolzlot** [sqlalchemy/sqlalchemy#10861] [Ideas]
I'm curious if "pipeline mode" will be at some point exposed or is it outside of this project scope.

I'm not sure if Postgres is the only database which has this feature.

More information about pipeline mode:

postgres docs:
https://www.postgresql.org/docs/current/libpq-pipeline-mode.html

psycopg3 docs:
https://www.psycopg.org/psycopg3/docs/advanced/pipeline.html

article about using pipeline mode in psycopg3:
https://blog.dalibo.com/2022/09/19/psycopg-pipeline-mode.html

asyncpg issue (not implemented fully yet, because doesn't allow manual pipeline mode):
https://github.com/MagicStack/asyncpg/issues/839

another discussion here which mentions it:
https://github.com/sqlalchemy/sqlalchemy/discussions/10270#discussioncomment-6804419

**Comment by zzzeek** [comment]
I would say that's cursor specific and you'd be using it against the DBAPI cursor:  https://docs.sqlalchemy.org/en/20/core/connections.html#working-with-the-dbapi-cursor-directly

SQLAlchemy's execution model is highly dependent on a very specific set of cursor behaviors, and if those things change, that would require an all-new execution model that's specific to the postgresql dialects.

**Reply by CaselIT** [reply]
maybe what could be done to better support things like these that require the use of specific driver functions is to have a better story for creating result object via public api (and maybe to compile the statements). May be mostly a documentation effort.

This would not require a vast change on the sqlalchemy internals, and users could better make use of the advanced features offered by the driver?

**Reply by zzzeek** [reply]
this use case does not appear to need a result object in the first place, it looks to be a way of emitting statements without getting any result at all.     What could be helpful here would be a way to compile a Core statement and run it on a cursor, but we pretty much have that with statement.compile()

**Reply by CaselIT** [reply]
I think that result return is supported, but it's not overly clear from the psycopg documentation.
I think you would use `nextset()` to get the results (at least you do so when using executemany of psycopg that uses pipelining and supports results)

**Reply by zzzeek** [reply]
I dont see any similarity with psycopg's pipeline API and anything else on any other DBAPI at all, so I dont see anything yet that could be generalized within SQLAlchemy other than, "here's the DBAPI connection"

**Reply by CaselIT** [reply]
indeed not, at most I think there could be a public api for generating results given a cursor and a statement/compiled

**Reply by MRigal** [reply]
Since psycopg 3.3, there is a even a https://www.psycopg.org/psycopg3/docs/api/cursors.html#psycopg.Cursor.results `results()` method as a better `nextset()` version.
The pipeline mode, eliminating the network overhead, would be a great improvement for our use-case

**Reply by zzzeek** [reply]
unfortunately https://github.com/sqlalchemy/sqlalchemy/discussions/10861#discussioncomment-8088361 still applies here
